### PR TITLE
update CI

### PR DESCRIPTION
**Pull request by krissetto**
- supersedes / closes https://github.com/docker/cli/pull/4793
- supersedes / closes https://github.com/docker/cli/pull/3128
- supersedes / closes https://github.com/docker/cli/pull/4822

**- What I did**
- Updated gha runners to  `ubuntu-22.04`
- Updated e2e test matrix to test against `23-dind`, `24-dind` and `25-dind` versions of the engine
- Small adjustments to some tests because of the engine upgrade (engine output text + cgroups checking)
- Temporarily skipping a problematic test (until a fix on moby is released)

**- How I did it**
magic

**- How to verify it**
CI tests are green

**- A picture of a cute animal (not mandatory but encouraged)**
TBD

### CODE REVIEW

**Comment by krissetto** [comment]
Commits will be squashed before merging

File: e2e/plugin/trust_test.go
Comment (krissetto) [inline e2e/plugin/trust_test.go:48]: this panic occurs [here](https://github.com/moby/moby/blob/master/plugin/fetch_linux.go#L203)

File: .github/workflows/build.yml
Comment (Bablzz) [inline .github/workflows/build.yml:22]: what do you think if use string/constant/global var :) for this? in future, for example, change only 1 var instead of every job:)

Comment (krissetto) [reply .github/workflows/build.yml:22]: I also thought about that, but iirc GitHub Actions wants a static value to be manually defined or else I'd expect it to be supported in the `defaults` keyword.  

There might be some ways to work around this, but I wouldn't block merging because of it :)

File: e2e/plugin/trust_test.go
Comment (krissetto) [reply e2e/plugin/trust_test.go]: [moby PR](https://github.com/moby/moby/pull/47299) related to this `docker plugin install` test that fails and is currently skipped

**Comment by laurazard** [comment]
(very WIP) PR for the skipped plugin test: https://github.com/moby/moby/pull/47299

**Comment by krissetto** [comment]
if merged we should be able to close this as well: https://github.com/docker/cli/pull/4793

**Comment by thaJeztah** [comment]
Oh! Forgot; and GitHub still doesn't allow reviewing commit messages;

Screenshot 2024-02-03 at 13 12 01


Could you:

- Update your sign-off (and probably git-config) to use your real name, not GitHub handle (see ["sign your work"](https://github.com/docker/cli/blob/v25.0.2/CONTRIBUTING.md#sign-your-work)) `Signed-off-by: Christopher Petito <......>`
- Make the first line of the commit message to be `< 72 chars` (if possible); try to make it 50 - 72 characters.
- In the commit body (empty line after first line, rest is considered "body", describe the other changes (could be a bullet-list), 72/80 chars line-length where possible.

Having the short title as first line makes sure some tooling/output doesn't break (GitHub itself also truncates after 72-chars, but other tooling may do similar).

And having a short summary in the body can sometimes help find back changes; for this PR _possibly_ we could break up things into logical commits (but I haven't verified);

- commit to fix the test for different versions of the engine
- commit to update the engine versions
- commit to update the Ubuntu version in GHA (I think this one depends on engine versions being updated for cgroups v2?)

**Comment by thaJeztah** [comment]
I need to update the repository settings before CI goes green, because the version of docker changed, and that's what's included in the "required checks" config (SIGH!)

Screenshot 2024-02-03 at 18 12 47

File: e2e/plugin/trust_test.go
Comment (thaJeztah) [reply e2e/plugin/trust_test.go:26]: This fix was backported to 23.0, 24.0 and 25.0, so we can probably un-skip this

File: e2e/testdata/Dockerfile.connhelper-ssh
Comment (krissetto) [reply e2e/testdata/Dockerfile.connhelper-ssh:14]: left a comment + link for context

File: e2e/compose-env.yaml
Comment (vvoland) [inline e2e/compose-env.yaml:7]: The default value `stable-dind` is still 19.03. We should probably default to the latest release?

File: e2e/testdata/Dockerfile.connhelper-ssh
Comment (krissetto) [reply e2e/testdata/Dockerfile.connhelper-ssh]: Set #syntax, default value for the build arg (and renamed it + added the the build arg docstring)

File: .github/workflows/e2e.yml
Comment (krissetto) [reply .github/workflows/e2e.yml]: Added 23-24-25, 19 needs some extra cgroupv2 testing after updating to ubuntu 22.04 runners

**Review by thaJeztah — COMMENTED** [review]
overall LGTM, thanks!

There's some comments I left after looking again; they may not be strictly blockers, but let me know if you want to work on those;

- The commit contains various changes, and not all of them are "equal"; would be nice to split (e.g.) the test-fixes to a separate commit. In this case it's probably "fine", but it's useful to have test-fixes separate (i.e., changed the test, and it still works before updating Ubuntu, and it continues to work after updating Ubuntu). In some of those cases, I also open a separate PR to verify those changes "in isolation", and stack the "update CI" changes on top (I usually rebase those once the first PR is merged).
- I left a comment for one of the tests; if I understand the change correctly, it's to account for cgroups v1 vs v2. While "v2" is the default on _current_ distros, v1 is not entirely gone (yet), so it would be nice if the test worked on _both_.
- I left a comment on the `t.Skip` - if it still fails even with the latest 25.0.3 engine, something is wrong with the fix. Ideally we would "un-skip" it for 25.0 though (we could use API version for determining if it's v25.0 or older), so that we have the coverage again, but let me know if that's complicated.

File: .github/workflows/e2e.yml
Comment (krissetto) [reply .github/workflows/e2e.yml:31]: iirc we cannot put runner versions in a matrix, it's not supported by gha, so we'd likely have to create a separate job altogether to have both the ubuntu versions tested

File: e2e/container/run_test.go
Comment (krissetto) [reply e2e/container/run_test.go:149]: Actually it doesn't seem to be specific to cgroupv2, but just a different way to verify cgroup functionality is available.  

I tested in this draft pr (https://github.com/docker/cli/pull/4867/files) splitting out only the tests as you suggested in the other comments. Tests are green on the older ubuntu (20.04) as well (makes me wonder, is cgroup v1/v2 really the issue here?)

File: e2e/plugin/trust_test.go
Comment (krissetto) [reply e2e/plugin/trust_test.go:26]: If [this PR](https://github.com/docker/cli/pull/4331) gets merged first, i'll steal the skip test helper and use it to target moby >= 25 (so skipping APIs older than 1.44 iirc)

File: e2e/container/run_test.go
Comment (thaJeztah) [reply e2e/container/run_test.go:149]: Hm... good one. Perhaps I assumed incorrectly that it was for cgroups v2 (I recalled there were some changes in this area from v1  -> v2, but perhaps I'm wrong and this wasn't one of them 😅) 🤔

File: .github/workflows/e2e.yml
Comment (thaJeztah) [reply .github/workflows/e2e.yml:31]: Ah, yeah, let's leave that for a future exercise. Don't think we need to make things too complicated for now. We could add a (todo?) comment to outline why, but not a blocker otherwise.

**Comment by thaJeztah** [comment]
Looks like you pulled in a merge-commit; can you try doing a rebase?

**Comment by krissetto** [comment]
Yeah I noticed 😓, i accidentally pressed the sync branch button on github but i was doing the rebase locally

**Comment by thaJeztah** [comment]
- https://github.com/docker/cli/pull/4873 was merged, so if you rebase, you can use that utility for the `t.Skip()` 👍

**Comment by krissetto** [comment]
@thaJeztah we'll have to remember to change these checks as well. Shall we use the lts+latest as the new e2e tests to check for?

> I need to update the repository settings before CI goes green, because the version of docker changed, and that's what's included in the "required checks" config (SIGH!)
> 
> Screenshot 2024-02-03 at 18 12 47

**Review by thaJeztah — APPROVED** [review]
LGTM, thanks!

(🙈 sorry, thought I already LGTM'd 😂)

**Pull request merged**
